EPF refactor

angaz opened the pull request:
Hi,

My EPF project is focused on the node-crawler.

I wanted to add a feature to check if a node is properly port-forwarded, basically what is described in #20, and I noticed the project needed some updates to work with recent versions of Go, and libraries.

I would say everything is basically changed, so I don't think reviewing each file would make sense. I tried to keep things as similar as possible, while still updating the libraries, and arranging the project similar to other Go projects.

What I've changed:
- Swapped sqlite library to a pure go version, so CGO isn't needed anymore
- Upgraded to the latest Ethereum library version, which also meant upgrading to urfave/cli/v2
- Moved to a single Go package, and single executable, with `crawler` and `api` sub-commands

Plans still to come in no particular order:
- [ ] Update the dockerfile and deployment instructions
- [ ] Some more refactoring. I feel the package layout is a bit strange. I think it can be made easier to understand.
- [ ] #20
- [ ] Fix some of the reported issues.
- [ ] Maybe: Bundle the frontend into the executable to make deployment easier (But compilation will require an extra step)

Please tell me if there's anything you would like me to look at or if you would like to discuss my project, my username is the same on Discord.

mohamedmansour commented:
Awesome thank you! Since we are gonna refactor it, are there any plans to refactor it so tests would be easier to create? For contributions this big, I would have recommended multiple pull requests, would be easier to follow.

angaz commented:
> Since we are gonna refactor it, are there any plans to refactor it so tests would be easier to create?

Yeah, this is something I was sort of thinking about. There are some places where I can see tests will be useful.

> For contributions this big, I would have recommended multiple pull requests, would be easier to follow.

Yeah I know. Just getting it to work with newer Go versions and updated libraries needed quite a few changes. The `urfave/cli` v1 to v2 was the most significant in this regard, as well as updating some things which are new/different in the geth code which is used.

It's still working now the way it did before I made any changes, at least as far as I can tell, there shouldn't be any functionality difference. Moving functionality into different packages really confused git quite a bit. It was mainly file renaming and function/type renaming. I think it should be pretty easy to see where a file was moved, but I am very freshly familiar with the code, so that would probably be a lot more difficult for someone who hasn't looked at it for some time. From what I remember doing, there should only be significant changes to how the API is run. Everything else should be pretty much how it was with a few minor tweaks. Maybe looking at each commit individually would shed some light on the process I took? They are not particularly well named. I didn't expect to make so many changes this quickly. We can even have a call on Discord if you think that would make it easier for me to explain it.

Would you like to review this before I shuffle things around in the packages some more?

mohamedmansour commented:
Hey @angaz i was focusing more on the Javascript stuff, and @MariusVanDerWijden quickly made the crawler. I reviewed it looks good. If you have a plan how you want to refactor it, please share some diagrams so we can have a better understanding where things should be. What I care about the most is testability, and improving or rewriting the parser because that is the most fragile part.

angaz commented:
> i was focusing more on the Javascript stuff, and @MariusVanDerWijden quickly made the crawler.

Ok, good to know. I have not looked at the front end part yet. I probably should have tried to run the whole system first. I guess there will also be some library upgrades there which would be nice to have. So it's something I can look at. It's not my area of expertise though. I've got mostly database/backend/ops knowledge. I've dabbled a bit with JavaScript/Typescript/React, so I should be able to find my way around the upgrading of the libraries.

Yeah the parser is also where I can see there are some problems. Possibly there needs to be a lot more logic there. So I will have a look at that as part of the "fixing issues" phase of the project.

I don't have a concrete plan for how I want to restructure things. I think the configuration of the crawler could be made more straight forward to follow or change the naming a bit. I think that's where I had some bad feelings and felt like I wanted to change something. It's still a very small amount of code, so I'm not too worried about how it will end up.

I think the shared sqlite database file is still OK. I'm not sure what are the deployment requirements within the EF setup. Would it be better if there was an external database? Say Postgres? Or is it still OK to have it as sqlite files shared between containers? I saw #5 I'm not sure exactly what is needed for this. Is there an example from another project that I could look at? With the single executable, there will be one fewer container to create, it will just have different starting commands/environment variables.

angaz commented on flake.nix:
```diff
@@ -0,0 +1,47 @@
+{
+  description = "Ethereum network crawler, API, and frontend";
```
This is part of a build system that I use: https://github.com/NixOS/nixpkgs. I didn't realize I committed the files. It makes it nicer for people who want to install and run the project, and for those who want to develop it because there is a shared development environment. It's up to you if you want to keep it. I will make it for my own use, and keep it uncommitted if it's not something you would like to have in the project.

I see there's also a project to run Ethereum projects with nix: https://github.com/nix-community/ethereum.nix
There's pretty cool. I should give it a shot. I never thought to look for it.

taxmeifyoucan replied:
+1 flake for nix deployment is very much appreciated!

angaz commented on pkg/crawlerdb/db.go:
```diff
@@ -1,23 +1,29 @@
 	"time"
 
-	_ "github.com/mattn/go-sqlite3"
+	_ "modernc.org/sqlite"
```
The new library is essentially a project where they have converted the C library into a pure Go one. This is so that we don't need the requirement for CGO. Just one less thing. The source code is here: https://gitlab.com/cznic/sqlite

angaz commented on pkg/vparser/vparser.go:
```diff
@@ -108,9 +108,9 @@ func parseVersion(input string) Version {
-		vers.Date = split[split_length - 1]
-		vers.Build = split[split_length - 2]
-		vers.Tag = strings.Join(split[1:split_length - 3], "")
+		vers.Date = split[split_length-1]
```
Go has an automatic formatter `go fmt` which should be run by all IDEs. But yes, we can add a linter if you would like to. https://github.com/golangci/golangci-lint is what I would use.

angaz commented on pkg/vparser/vparser.go:
```diff
@@ -1,4 +1,4 @@
-package parser
+package vparser
```
The package name just being `parser` felt a bit too generic. I saw one file was named `vparser.go` and thought that was good enough. We could use `versionparser`. I thought `vparser` was a good middle ground.

angaz commented on cmd/crawler/main.go:
```diff
@@ -40,19 +41,21 @@ func init() {
-		fmt.Fprintf(os.Stderr, "No such command: %s\n", cmd)
-		os.Exit(1)
-	}
+	// app.CommandNotFound = func(ctx *cli.Context, cmd string) {
```
I'm not sure why it was here. The "command not found" prints perfectly well without it, so I wanted to give it a try just using the default. It's surely not a good idea to have commented out code committed like this. So I will remove it.

angaz commented:
I have a few more commits to refactor the `docker compose` and I updated the frontend libraries to work with node 18, and updated the README accordingly. Those commits will come soon. I don't believe it will be too much work to review them. They are pretty simple and I think it will be good to have the docker setup and docs updated in the same PR.

angaz commented:
https://github.com/ethereum/node-crawler/pull/40/files/ab1a81df5ab929b643f201bfb6f4600473826856..56b4df2c60c498d09f0072d4af8eb86a9ba18cd6

Should be the new commits I spoke about. I have no more planned changes. The new things will come in future PRs. I think this is good to go.

With the updated `docker compose`, I believe it should be easier to deploy. I see the current https://pscrawl1.ethdevops.io/ seems to be broken. I tested the new `docker compose` on my machine and it seems like it's working as expected.

angaz commented:
I will rebase when we are happy with everything. I see there is a merge conflict with one of the Dockerfiles.

mohamedmansour commented:
> I will rebase when we are happy with everything. I see there is a merge conflict with one of the Dockerfiles.

Let me know when you are ready to merge and I will merge

angaz commented:
Great. I will be home in an hour or two and I'll resolve the merge conflict :rocket:

angaz commented:
@mohamedmansour it's all ready to go :rocket:

mohamedmansour commented:
Awesome, great work!

MariusVanDerWijden commented:
Hey @angaz sorry I didn't have time to review your PR, I took a quick peek now and it looks good. If you have some time it would be really nice to hop on a quick a call to discuss your changes and if you have some more ideas to improve the crawler
